### ISSUE TITLE
Casting a hexadecimal literal with MSB set to a byte

### ISSUE THREAD

**Issue by Maxi17** [typetools/checker-framework#2367]
Consider this snippet:

```
public class Issue2367 {
    byte b1 = 0x4B;
    byte b3 = (byte) 0x4B;
    byte b4 = (byte) 0x8B;   // false positive warning on this line
}
```

Running javac -processor index Issue2367.java produces a warning:
```
Warning:(4, 14) java: [cast.unsafe] cast from "@IntVal(139) int" to "@IntVal(-117) byte" cannot be statically verified
```
Normally, this shouldn't be a problem, as the value 0x8B is 8 bits long (10001011), so the cast shouldn't issue any warning. I think this happens because 0x8B is a negative value when taken as a byte and positive when taken as an int, so the checker believes there is something wrong. There is no warning when the hexadecimal number is 0x4B (01001011),  the warning disappears. Therefore, the warning is issued because the hexadecimal value is always considered to be an integer by Java, so changing its intVal by casting it to a byte produces a warning.

**Comment by mernst** [comment]
I committed a test case as `framework/tests/value/Issue2367.java`.

Run it like this:
```checker/bin/javac -processor org.checkerframework.common.value.ValueChecker framework/tests/value/Issue2367.java```

**Comment by mernst** [comment]
Consider the following example.
```
public class Issue2367 {
    byte b4 = (byte) 139;  // b4 == -117
    byte b7 = (byte) 0x8B; // 0x8B == 137, and b4 == -117
}
```

(We need the `(byte)` cast because otherwise javac issues an error.

A warning is desirable for the first line, because it changes a value from 139 to -117.  This warning is analogous to the javac error that the cast suppressed.
The second example is identical Java (the same bytecode is generated).  Your feature request is to not issue any warning, based on the difference in how the source code was written.

This is a reasonable feature request (albeit a bit counterintuitive at first).

Unfortunately, I don't see a way to implement it.  The program element "(byte) 0x8B" has already been converted to "(byte)139" by javac before the Checker Framework gets access to it.  In other words, the Checker Framework cannot tell the difference between the program text `(byte) 139` and `(byte) 0x8B`, in order to make the distinction that you want.

Here are some workarounds:
 * You could suppress the warning.
 * You could use the Signedness Checker.  (I have not tried doing this on your example yet.)
 * A feature could be added to the Constant Value Checker to only issue warnings at casts if the bit representation would change, even if the value would change.  (This wouldn't be on by default.)
